src/AspNetCore/Src/applicationinfo.cpp

```diff
@@ -330,3 +321,115 @@ APPLICATION_INFO::FindNativeAssemblyFromGlobalLocation(STRU* struFilename)
 Finished:
     return hr;
 }
+
+// 
+// Tries to find aspnetcorerh.dll from the application
+// Calls into hostfxr.dll to find it.
+// Will leave hostfxr.dll loaded as it will be used again to call hostfxr_main.
+// 
+
+HRESULT
+APPLICATION_INFO::FindNativeAssemblyFromHostfxr(
+    STRU* struFilename
+)
+{
+    HRESULT     hr = S_OK;
+    HANDLE      hNativeRequestHandler = INVALID_HANDLE_VALUE;
+    STRU        struApplicationFullPath;
+    STRU        struNativeSearchPaths;
+    STRU        struNativeDllLocation;
+    HMODULE     hmHostFxrDll = NULL;
+    INT         intHostFxrExitCode = 0;
+    INT         intIndex = -1;
+    INT         intPrevIndex = 0;
+    BOOL        fFound = FALSE;
+    const DWORD BUFFER_SIZE = 1024 * 10;
+    WCHAR       pwszNativeSearchPathsBuffer[BUFFER_SIZE];
+    
+    DBG_ASSERT(struFileName != NULL);
+    DBG_ASSERT(pHostFxrParameters != NULL);
+
+    hmHostFxrDll = LoadLibraryW(m_pConfiguration->QueryHostFxrLocation()->QueryStr());
+    
+    if (hmHostFxrDll == NULL)
+    {
+        // Could not load hostfxr
+        goto Finished;
+    }
+
+    hostfxr_get_native_search_directories_fn pFnHostFxrSearchDirectories = (hostfxr_get_native_search_directories_fn)
+        GetProcAddress(hmHostFxrDll, "hostfxr_get_native_search_directories");
+
+    if (pFnHostFxrSearchDirectories == NULL)
+    {
+        // Host fxr version is incorrect
+        hr = E_FAIL;
+        goto Finished;
+    }
+
+    intHostFxrExitCode = pFnHostFxrSearchDirectories(
+        *m_pConfiguration->QueryHostFxrArgCount(),
+        *m_pConfiguration->QueryHostFxrArguments(),
+        pwszNativeSearchPathsBuffer, 
+        BUFFER_SIZE
+    );
+
+    if (intHostFxrExitCode != 0)
+    {
+        // Call to hostfxr failed.
+        hr = E_FAIL;
+        goto Finished;
+    }
+
+    if (FAILED(hr = struNativeSearchPaths.Copy(pwszNativeSearchPathsBuffer)))
+    {
+        goto Finished;
+    }
+    
+    // The native search directories are semicolon delimited.
+    // Split on semicolons, append aspnetcorerh.dll, and check if the file exists.
+    while ((intIndex = struNativeSearchPaths.IndexOf(L";", intPrevIndex)) != -1)
+    {
+        if (FAILED(hr = struNativeDllLocation.Copy(struNativeSearchPaths.QueryStr(), intIndex - intPrevIndex)))
+        {
+            goto Finished;
+        }
+
+        if (!struNativeDllLocation.EndsWith(L"\\"))
+        {
+            if (FAILED(hr = struNativeDllLocation.Append(L"\\")))
+            {
+                goto Finished;
+            }
+        }
+
+        if (FAILED(struNativeDllLocation.Append(g_pwzAspnetcoreRequestHandlerName)))
```

Comment:
set the hr value from the append

Reply:
Also, with these different failure modes, do we want to log causes as well? I'd like it if it was super straightforward to diagnose common mistakes and recover from them.

Reply:
For sure. I think we are batching all logging related things into the same PR.